### DISCUSSION TITLE
Integrate Casbin and FastAPI to support access control

### DISCUSSION THREAD

**Discussion by shizidushu** [fastapi/fastapi#7831] [Questions]
**Is your feature request related to a problem? Please describe.**
It is a useful addition to have access control over FastAPI. Some framework have builtin support and some can get this feature with a plugin.

**Describe the solution you'd like**
According to its docs, [Casbin](https://casbin.org/en/) provides support for enforcing authorization based on various access control models.
Here is the function I image:
- Check permission before handle the request. Maybe implement is using a custom class APIRoute class.
- Sync FastAPI path operation with Casbin policy file (or policy storage adapter).
  Maybe a function to execute on application startup; it gets all path operation's path, method and some other optional information like tags, user_id and user_role, then write it to Casbin policy file.
  It cleans redundant row in Casbin policy file and keep it update with FastAPI routes.
- Document to describe how to use it (with example code).

**Comment by prostomarkeloff** [comment]
I think it should be implemented in independent library, not in fastapi core.

**Comment by NJannasch** [comment]
This could be easily achived with a custom middleware.

Dummy Code Example:
```
from starlette.datastructures import Headers
from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.status import HTTP_401_UNAUTHORIZED
from starlette.types import ASGIApp, Receive, Scope, Send

class CasbinMiddleware:
    def __init__(self, app: ASGIApp, enforcer) -> None:
        self.app = app
        self.enforcer = enforcer

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:

        # ToDo Adjust me
        headers = Headers(scope=scope)
        token = headers.get('authorization', None)
        # validate user and get user from token
        user = 'test123'
        #######################################

        request = Request(scope, receive)

        if self.check_permission(user=user, request=request):
            await self.app(scope, receive, send)
            return

        # Invalid request
        response = JSONResponse(
            status_code=HTTP_401_UNAUTHORIZED,
            content="Insufficient permissions",
            headers={"WWW-Authenticate": "Bearer"},
        )

        await response(scope, receive, send)
        return

    def check_permission(self, user: str, request: Request):
        path = request.path
        method = request.method
        return self.enforcer.enforce(user, path, method)
```


In your main file:
```
from middleware import CasbinMiddleware
import casbin
...
enforcer = casbin.Enforcer("authz_model.conf", "authz_policy.csv")
app.add_middleware(CasbinMiddleware,enforcer=enforcer)
```

**Comment by shizidushu** [comment]
@NJannasch Thanks for your reply. Although I have implemented some function I imaged some days ago. I check the permissions with `Depends` to get more control. Here is a piece of code:
```python
def get_current_active_authorized_user(request: Request, current_active_user: User = Security(get_current_active_user)) -> Optional[User]:
    username = current_active_user.Username
    path = request.state.path

    sub = username
    obj = path
    act = request.method.lower()

    print(sub, obj, act)

    if enforcer.enforce(sub, obj, act):
        return current_active_user
    else:
        raise HTTPException(status_code = HTTP_403_FORBIDDEN, detail="No permission")
```

I just make all my code public here: https://github.com/shizidushu/simple-report-data-table-vuetify

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks @shizidushu for reporting back and closing the issue :+1:

**Comment by Neko1313** [comment]
Guys, I made a small library for the integration of fastapi and casbin, which makes it possible to add a rights check to the router levels. Could you evaluate it?

https://pypi.org/project/casbin-fastapi-decorator/
